src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -4200,6 +4203,38 @@ private IOperation ConvertToIDisposable(IOperation operand, ITypeSymbol iDisposa
         {
             StartVisitingStatement(operation);
 
+            // `lock (l) { }` on value of type `System.Threading.Lock` is lowered to `using (l.EnterLockScope()) { }`.
+            if (operation.LockedValue.Type?.IsLockType() == true)
+            {
+                var lockTypeInfo = operation.LockedValue.Type.TryFindLockTypeInfo();
+
+                var lockObject = VisitRequired(operation.LockedValue);
+
+                var enterLockScope = lockTypeInfo.EnterLockScopeMethod is { } targetMethod
+                    ? new InvocationOperation(
+                        targetMethod: targetMethod,
+                        constrainedToType: null,
+                        instance: lockObject,
+                        isVirtual: targetMethod.IsVirtual || targetMethod.IsAbstract || targetMethod.IsOverride,
+                        arguments: ImmutableArray<IArgumentOperation>.Empty,
+                        semanticModel: null,
+                        syntax: lockObject.Syntax,
+                        type: targetMethod.ReturnType,
+                        isImplicit: true)
+                    : MakeInvalidOperation(type: null, lockObject);
```

Comment:
>MakeInvalidOperation(type: null, lockObject) [](http://example.com/codeflow?start=22&length=44)

I suggest simplifying an error recovery for `Lock` object. If any of the required APIs are missing, we visit ```operation.LockedValue``` the way ```VisitExpressionStatement``` visits its underlying expression, then we wrap the result with ```MakeInvalidOperation``` and add an expression statement with that invalid operation (again, as '''VisitExpressionStatement''' does that. Then we simply do ```VisitStatement(operation.Body)```.